osu.Game/Beatmaps/Formats/LegacyBeatmapEncoder.cs

```diff
@@ -230,14 +230,47 @@ void outputControlPointAt(double time, bool isTimingPoint)
                 if (effectPoint.OmitFirstBarLine)
                     effectFlags |= LegacyEffectFlags.OmitFirstBarLine;
 
-                writer.Write(FormattableString.Invariant($"{(int)beatmap.ControlPointInfo.TimingPointAt(time).TimeSignature},"));
+                writer.Write(FormattableString.Invariant($"{(int)legacyControlPoints.TimingPointAt(time).TimeSignature},"));
                 writer.Write(FormattableString.Invariant($"{(int)toLegacySampleBank(tempHitSample.Bank)},"));
                 writer.Write(FormattableString.Invariant($"{toLegacyCustomSampleBank(tempHitSample)},"));
                 writer.Write(FormattableString.Invariant($"{tempHitSample.Volume},"));
                 writer.Write(FormattableString.Invariant($"{(isTimingPoint ? '1' : '0')},"));
                 writer.Write(FormattableString.Invariant($"{(int)effectFlags}"));
                 writer.WriteLine();
             }
+
+            void extractDifficultyControlPoints(HitObject hitObject)
+            {
+                if (!isOsuRuleset)
+                    return;
+
+                var hDifficultyPoint = hitObject.DifficultyControlPoint;
+
+                if (!hDifficultyPoint.IsRedundant(lastRelevantDifficultyPoint))
+                {
+                    legacyControlPoints.Add(hDifficultyPoint.Time, hDifficultyPoint);
+                    lastRelevantDifficultyPoint = hDifficultyPoint;
+                }
+
+                foreach (var nested in hitObject.NestedHitObjects)
+                    extractDifficultyControlPoints(nested);
+            }
+
+            void extractSampleControlPoints(HitObject hitObject)
+            {
+                // process nested objects first as these use EndTime for legacy storage.
```

Comment:
not sure this comment is clear enough. basically the sample points need to be processed with nested first to avoid nested sample data being lost due to the presence of the hitobject itself. reversing the order makes tests fail which is how i got here.